File: ios/Sources/RolloutBucket.swift
import CryptoKit
import Foundation

public enum RolloutBucket {
    public static func calculate(key: String, bucketingId: String) -> Int {
        let digest = SHA256.hash(data: Data("\(key)\0\(bucketingId)".utf8))
        let prefix = digest.prefix(4).reduce(UInt32.zero) { value, byte in
            (value << 8) | UInt32(byte)
        }
        return Int(prefix % 10_000)
    }
}

File: android/src/main/kotlin/dev/pam/featureflags/RolloutBucket.kt
package dev.pam.featureflags

import java.security.MessageDigest

object RolloutBucket {
    fun calculate(key: String, bucketingId: String): Int {
        val digest = MessageDigest.getInstance("SHA-256")
            .digest("$key\u0000$bucketingId".toByteArray(Charsets.UTF_8))
        val prefix = ((digest[0].toLong() and 0xff) shl 24) or
            ((digest[1].toLong() and 0xff) shl 16) or
            ((digest[2].toLong() and 0xff) shl 8) or
            (digest[3].toLong() and 0xff)
        return (prefix % 10_000L).toInt()
    }
}
